cmd/signature-v4-utils.go

```diff
@@ -57,7 +57,25 @@ func skipContentSha256Cksum(r *http.Request) bool {
 
 	// If x-amz-content-sha256 is set and the value is not
 	// 'UNSIGNED-PAYLOAD' we should validate the content sha256.
-	return !(ok && v[0] != unsignedPayload)
+	if ok {
+		switch v[0] {
+		case unsignedPayload:
+			return true
+		case emptySHA256:
+			// some broken clients set empty-sha256
+			// with > 0 content-length in the body,
+			// we should skip such clients and allow
+			// blindly such insecure clients.
+			if r.ContentLength > 0 && !globalCLIContext.StrictS3Compat {
+				// We return true only in situations when
+				// deployment has asked MinIO to allow for
+				// such broken clients.
+				return true
+			}
+		}
+		return false
+	}
+	return true
```

Comment:
Simplify and skip check if length is actually 0.

```suggestion
	// Skip if no header was set.
	if !ok {
		return true
	}
	// If x-amz-content-sha256 is set and the value is not
	// 'UNSIGNED-PAYLOAD' we should validate the content sha256.
	switch v[0] {
	case unsignedPayload:
		return true
	case emptySHA256:
		// some broken clients set empty-sha256
		// with > 0 content-length in the body,
		// we should skip such clients and allow
		// blindly such insecure clients.

		// We return true only in situations when
		// deployment has asked MinIO to allow for
		// such broken clients, or if length is actually 0.
		return r.ContentLength == 0 || !globalCLIContext.StrictS3Compat
	}
	return false
```

Reply:
@klauspost I wrote it this way since it's intricate to show that it's more readable - I can change it to the new style as well

Reply:
NOTE: we really don't want to skip content-sha256 in any situation even with contentlength == 0 we want to skip only when contentlength > 0 if the incoming value is `emptySha256`